### DISCUSSION TITLE
[Bug]:Tables not rendering alignment according to spec

### DISCUSSION THREAD

**Discussion by yengip** [django-wiki/django-wiki#1319] [Bug] [answered]
### Describe the bug

Using the colon notation in markdown tables does not seem to render align attributes per the spec.
|:---|:---:|---:|
I thought it might be the version of wiki or markdown I was using but I've tried wiki v0.7.9 with markdown 2.2 and wiki v0.10 with markdown 3.3 with the same results.  I have tried the markdown 2.2 module in a python shell, using the extensions=["tables"] argument, and that renders the align attributes correctly.  So I'm left thinking that something in the wiki code is stopping the attributes getting through.  My knowledge of python isn't up to tracing the code to find out what might be happening to prevent the attributes finding their way into the final rendered html.
Any pointers?

### Steps to reproduce.

1. Create an article with the following text:
Table  
| header1 | header 2 | header 3 |
|:-----|:-----:|-----:|
| left | ctr | right |

3. Save it
4. View the table in the html page (no alignment)
5. View the source, look for the <td> tags - no align="left" or "center" or "right"

### Expected behavior

The rendered html should include the correct align attributes in the <td> tags

### Python Version

3.9.6

### Django Version

3.2.6

**Comment by benjaoming** [comment, marked as answer]
> View the source, look for the tags - no align="left" or "center" or "right"

True. Actually, I think these need to be added to the _allow list_ of the bleach library that sanitizes all the Markdown HTML outputs. I wonder if the intention is to use these tags, though. It might be more conventional to add a CSS class.

**Reply by yengip** [reply]
Thanks for the pointer, you're exactly right.  I added it to this section of sanitize.py and now it works.
```#: Map of allowed attributes by tag  
ALLOWED_ATTRIBUTES = {
    "a": ["href", "title"],
    "abbr": ["title"],
    "acronym": ["title"],
    "td": ["align"],
}
```

I should probably try and work out how to submit a pull request, right?

**Reply by oscarmcm** [reply]
I was wondering the same, should we support that? I think so, as it get us more closer to the markdown spec. So yeah, feel free to submit a new PR.

**Reply by yengip** [reply]
Never done that before so it'll take me a while to learn how.  Google here I come :-)

**Reply by yengip** [reply]
Meanwhile, is there a mechanism for me to override the current bleach setting by doing something in my project settings.py, or am I stuck editing the original module code?

**Reply by yengip** [reply]
Also, deeply appreciate you guys rapid response and insight.  Legends.

**Reply by benjaoming** [reply]
I guess that we would want the tables extension to use the `align` attribute even though it's considered legacy in html5. However, the alternative to use `style` isn't possible since we don't want to allow random `style` attributes and CSS rules on any element.

See also: https://python-markdown.github.io/extensions/tables/

So yeah, when Markdown tables expect the `align` attribute to be defined, I think that's how we should go about with it :+1: I think that it might even be as a default setting, since it can't cause much harm.

**Reply by benjaoming** [reply]
@yengip as @oscarmcm noted [below](https://github.com/django-wiki/django-wiki/discussions/1319#discussioncomment-7386405), then you can already fix your issue with a local setting for your project.

**Comment by oscarmcm** [comment]
> Never done that before so it'll take me a while to learn how. Google here I come :-)

Some resources:

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request

https://github.com/openpnp/openpnp/wiki/Your-First-Pull-Request

**Reply by yengip** [reply]
Thanks, I've submitted a pull request.  It's a one line change.  I checked it in my own project dev environment and it worked.  
I cloned the repo but struggled to get it working so I could run all the tests and maybe make a specific test case for it.  Got a lot to learn :-(

**Reply by benjaoming** [reply]
The tests are also run in CI and judging from your PR, it works :tada:

**Reply by benjaoming** [reply]
A thing you can do to improve the PR would be to add a test case for tables and tables with aligned data.

You could add a test case that asserts that generated HTML contains the expected alignment tag, similar to this test case:

https://github.com/django-wiki/django-wiki/blob/main/tests/core/test_markdown.py#L44

**Reply by yengip** [reply]
Thanks, I wanted to do that but I wasn't able to get the whole thing working in my local environment so I could actually run the tests.  I'm sure I'll figure it out eventually but this is my first time cloning a package repo and trying to work on it.  Do I need to clone the repo into my virtual env structure over the top of an installed version (so that pip has handled all the dependencies) and then put the test project and the tests somewhere else?  Haven't found any tutorial/guide info on how to do this as yet.  I'll stop posting here and ask on stackoverflow.

**Reply by oscarmcm** [reply]
@yengip 

https://django-wiki.readthedocs.io/en/main/development/index.html

Hope this helps :)

**Reply by yengip** [reply]
It does help a bit, thanks.  Never seen `hatch` before so that's another thing to learn about.  My IDE of choice for django is VSCode, but I can't find a way to get it to play nicely with `hatch`.  It seems to want a nice static venv and I haven't worked out how to get it to handle linking the working directory with the `hatch` env.

**Reply by oscarmcm** [reply]
I stopped using VSCode long time ago, so I'm not sure how it picks the environment, what I might suggest is to look where `hatch` saves the environment and then set that path the VSCode config.

https://hatch.pypa.io/latest/environment/

**Comment by oscarmcm** [comment]
> Meanwhile, is there a mechanism for me to override the current bleach setting by doing something in my project settings.py, or am I stuck editing the original module code?

Yeah, see here https://github.com/django-wiki/django-wiki/blob/main/src/wiki/conf/settings.py#L104-L106

Basically in your local settings just set `WIKI_MARKDOWN_HTML_ATTRIBUTES` to the new values.

**Reply by yengip** [reply]
Thanks, much appreciated.  Works a treat.

**Reply by yengip** [reply]
Just a follow up thought.  This code in src/wiki/conf/settings.py...
```
MARKDOWN_HTML_ATTRIBUTES.update(
    getattr(django_settings, "WIKI_MARKDOWN_HTML_ATTRIBUTES", {})
)
```
...has the effect of overwriting tag entries in `MARKDOWN_HTML_ATTRIBUTES` with any that are in my project settings, right?  
E.g. my `WIKI_MARKDOWN_HTML_ATTRIBUTES = {'td': ['align']}` will completely replace the carefully constructed default entry `MARKDOWN_HTML_ATTRIBUTES = {..., 'td': {'class', 'id', 'rel', 'target'}...}`.

Is this the desired behaviour?  It does give me, as the project developer, complete control of the attribute whitelist.  But it also requires me to delve into the library code to find out what the default setting is if I don't want to destroy it with my own attempt.  If I haven't done that investigation, I would probably assume that I'm adding to the attribute set rather than overwriting it (_this was actually my own assumption at first_).  That might have unintended consequences caused by having those default attributes suddenly wiped out by `bleach`.

Is it worth considering using a list comprehension to update each tag's attribute list instead of replacing it?
`MARKDOWN_HTML_ATTRIBUTES = {k:list(set(v + WIKI_MARKDOWN_HTML_ATTRIBUTES[k])) if k in WIKI_MARKDOWN_HTML_ATTRIBUTES else v for k,v in MARKDOWN_HTML_ATTRIBUTES.items()}`
It's a bit of a mouthful but it seems to do the trick.  Might be more readable using a for loop.

**Reply by benjaoming** [reply]
> Is this the desired behaviour?

I think that it's okay, yes. It's normal that once a user defines a setting in their project, then they take responsibility of correctly defining the entire setting.

Magically extending things has the side effect of making it impossible to remove stuff.

**Reply by oscarmcm** [reply]
As @benjaoming stated, that's the desired behavior:

```py
Python 3.10.2 (main, Feb 24 2022, 15:33:09) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> a = {'a': 12, 'b': 33}
>>> a
{'a': 12, 'b': 33}
>>> a.update({'c': 55})
>>> a
{'a': 12, 'b': 33, 'c': 55}
>>> a.update({'b': 25})
>>> a
{'a': 12, 'b': 25, 'c': 55}
>>>
```
